### PR TITLE
Add URI pattern matching to wrap-apm-transaction

### PR DESCRIPTION
**Pull request by jarcane**
This adds a new feature to the `wrap-apm-transaction` middleware, so that you can give it a list of URI patterns to filter and group transactions better.

One open question is whether the matching should be eager, as it is here, or stricter (exact pattern match only). Making it eager did seem to fix some weird issues with trailing slashes possibly? But that's something that can probably be fixed.

### CODE REVIEW

**Review by AriPaaWun — COMMENTED** [review]
Nice documentation 👍 
How about some tests?

**Comment by jarcane** [comment]
> Nice documentation 👍
> How about some tests?

I had the same thought. Want to make sure to nail down any weird edge cases. Have had some odd results with trailing slashes.

File: README.md [hunk 1]
```diff
@@ -170,6 +170,33 @@ For convenience, the `clojure-elastic-apm.ring` namespace provides `wrap-apm-tra
...
+The `wrap-apm-transaction` middleware takes an optional argument, that allows you to pass a list of URI patterns, which 
+will be matched against any requests, and only log a transaction for those that match the pattern:
+
+```clojure
+(require '[clojure-elastic-apm.ring :as apm-ring])
+
+(def app (->> app-routes
+              wrap-params
+              wrap-json-params
+              apm-ring/wrap-apm-transaction ["/*/*/_"]))
```
Comment (jarcane) [inline README.md]: missing parens here

File: src/clojure_elastic_apm/ring.clj [hunk 2]
```diff
@@ -2,11 +2,45 @@
...
+           (when status
+             (.setResult tx (str "HTTP " status)))
+           response)))))
+  ([patterns handler]
+   (fn [{:keys [request-method uri] :as request}]
+     (let [matched (->> patterns
+                        (map #(match-uris % uri))
+                        (drop-while false?)
+                        (first))
+           short-uri (if matched matched "_")
```
Comment (jarcane) [inline src/clojure_elastic_apm/ring.clj]: This line isn't necessary; we don't use the URI anywhere if it doesn't match a pattern, since we don't like a transaction

File: README.md [hunk 3]
```diff
@@ -170,6 +170,33 @@ For convenience, the `clojure-elastic-apm.ring` namespace provides `wrap-apm-tra
...
+
+(def app (->> app-routes
+              wrap-params
+              wrap-json-params
+              (apm-ring/wrap-apm-transaction ["/*/*/_"])))
+```
+
+The pattern format breaks up the URI into segments on the slashes ("/") and matches the segments with the following syntax:
+- `*`: matches any data in the segment, and keeps it in the resulting transaction name
+- `_`: matches any data in the segment, but ignores it in the resulting tx name, instead inserting a splat('*') character
```
Comment (keveri) [inline README.md]: Should `splat` be `asterisk`? :male_detective:  :policewoman:

**Review by AriPaaWun — APPROVED** [review]
:+1:

**Review by keveri — APPROVED** [review]
Looks good :+1: :gem:

File: src/clojure_elastic_apm/ring.clj [hunk 4]
```diff
@@ -2,11 +2,44 @@
...
+                  (cond
+                    (= p "*") u
+                    (= p "_") "*"
+                    (= p u) u
+                    :else false))]
+    (if (> (count pattern-segs)
+           (count uri-segs))
+      false
+      (let [matches (map matcher pattern-segs uri-segs)
+            matched? (reduce #(and %1 %2) matches)]
```
Comment (keveri) [inline src/clojure_elastic_apm/ring.clj:18]: `kibit` suggests the following:
```
Consider using:
  and
instead of:
  #(and %1 %2)
```

File: README.md (hunk 3 above)
Comment (jarcane) [reply README.md]: Either is correct, though asterisk may be more clear and less vernacular.

File: src/clojure_elastic_apm/ring.clj (hunk 4 above)
Comment (jarcane) [reply src/clojure_elastic_apm/ring.clj:18]: Kibit is wrong here. :D `and` is a macro, you can't pass it on its own as if it were a function.
